### DISCUSSION TITLE
Server unable to connect to docker container on windows?

### DISCUSSION THREAD

**Discussion by bteodoru8** [serverpod/serverpod#2007] [Q&A] [answered]
Hello, I have recently started working with serverpod. A week ago I played around with authentication and everything was working fine, however, today I was trying to start a new project, but I am unable to run the server on windows. Here you can see the results from my console.

```
PS D:\Dev\****\project\project_server> docker compose up --build --detach
[+] Running 2/2
 ✔ Container project
_server-redis-1     Started                                                                      0.0s
 ✔ Container project
_server-postgres-1  Started                                                                      0.0s
PS D:\Dev\****\project\project_server> dart bin/main.dart --apply-migrations --role=maintenance
/C:/Users/lxxma/AppData/Local/Pub/Cache/hosted/pub.dev/postgres-2.6.4/lib/src/substituter.dart:24:52: Warning: Operand of null-aware operation '!' has type 'Map<String, dynamic>' which excludes null.
 - 'Map' is from 'dart:core'.
    replace ??= (spec, index) => converter.convert(values![spec.name]);
                                                   ^
SERVERPOD version: 1.2.5, dart: 3.3.0 (stable) (Tue Feb 13 10:25:19 2024 +0000) on "windows_x64", time: 2024-03-12 15:04:14.937094Z
mode: development, role: maintenance, logging: normal, serverId: default
Failed to connect to the database. Retrying in 10 seconds. TimeoutException after 0:00:15.000000: Future not completed
===== asynchronous gap ===========================
package:postgres_pool/postgres_pool.dart 351:18               PgPool.run.<fn>
package:retry/retry.dart 131:16                               RetryOptions.retry
package:postgres_pool/postgres_pool.dart 349:14               PgPool.run
package:serverpod/src/database/database_connection.dart 42:5  DatabaseConnection.testConnection
package:serverpod/src/server/serverpod.dart 673:9             Serverpod._connectToDatabase
package:serverpod/src/server/serverpod.dart 354:19            Serverpod.start.<fn>
package:serverpod/src/server/serverpod.dart 341:5             Serverpod.start
package:project_server/server.dart 33:3                         run

Database configuration:
database host: localhost
database port: 8090
database name: project
database user: postgres
database require SSL: false
database unix socket: false
database pass: ********

Failed to connect to the database. Retrying in 10 seconds. TimeoutException after 0:00:15.000000: Future not completed
===== asynchronous gap ===========================
package:postgres_pool/postgres_pool.dart 351:18               PgPool.run.<fn>
package:retry/retry.dart 131:16                               RetryOptions.retry
package:postgres_pool/postgres_pool.dart 349:14               PgPool.run
package:serverpod/src/database/database_connection.dart 42:5  DatabaseConnection.testConnection
package:serverpod/src/server/serverpod.dart 673:9             Serverpod._connectToDatabase
package:serverpod/src/server/serverpod.dart 354:19            Serverpod.start.<fn>
package:serverpod/src/server/serverpod.dart 341:5             Serverpod.start
package:project_server/server.dart 33:3                         run

Failed to connect to the database. Retrying in 10 seconds. TimeoutException after 0:00:15.000000: Future not completed
===== asynchronous gap ===========================
package:postgres_pool/postgres_pool.dart 351:18               PgPool.run.<fn>
package:retry/retry.dart 131:16                               RetryOptions.retry
package:postgres_pool/postgres_pool.dart 349:14               PgPool.run
package:serverpod/src/database/database_connection.dart 42:5  DatabaseConnection.testConnection
package:serverpod/src/server/serverpod.dart 673:9             Serverpod._connectToDatabase
package:serverpod/src/server/serverpod.dart 354:19            Serverpod.start.<fn>
package:serverpod/src/server/serverpod.dart 341:5             Serverpod.start
package:project_server/server.dart 33:3                         run

Failed to connect to the database. Retrying in 10 seconds. TimeoutException after 0:00:15.000000: Future not completed
===== asynchronous gap ===========================
package:postgres_pool/postgres_pool.dart 351:18               PgPool.run.<fn>
package:retry/retry.dart 131:16                               RetryOptions.retry
package:postgres_pool/postgres_pool.dart 349:14               PgPool.run
package:serverpod/src/database/database_connection.dart 42:5  DatabaseConnection.testConnection
package:serverpod/src/server/serverpod.dart 673:9             Serverpod._connectToDatabase
package:serverpod/src/server/serverpod.dart 354:19            Serverpod.start.<fn>
package:serverpod/src/server/serverpod.dart 341:5             Serverpod.start
package:project_server/server.dart 33:3                         run

Failed to connect to the database. Retrying in 10 seconds. TimeoutException after 0:00:15.000000: Future not completed
===== asynchronous gap ===========================
package:postgres_pool/postgres_pool.dart 351:18               PgPool.run.<fn>
package:retry/retry.dart 131:16                               RetryOptions.retry
package:postgres_pool/postgres_pool.dart 349:14               PgPool.run
package:serverpod/src/database/database_connection.dart 42:5  DatabaseConnection.testConnection
package:serverpod/src/server/serverpod.dart 673:9             Serverpod._connectToDatabase
package:serverpod/src/server/serverpod.dart 354:19            Serverpod.start.<fn>
package:serverpod/src/server/serverpod.dart 341:5             Serverpod.start
package:project_server/server.dart 33:3                         run
```

And it keeps repeating indefinitely.

I use windows on my main machine, but I have a macbook too, so I decided to set up the project on macbook first, where it ran perfectly. I uploaded the exact same project in a repo, cloned it on my windows machine and it has the same issue. I tried running the demo I wrote last week, which worked perfectly fine but I was getting the exact same results. Looked up a solution, however it seems there was no one with the same problem. I even reinstalled docker and ran everything from 0, to no success. I had tried the solution in #930 as well, but to no avail. It might be a stupid issue on my end, but I kindly ask you to help me if possible as I really like serverpod, Also my team would like to start using it as our backend framework for our future apps.

Looking forward to your cooperation!

**Comment by bteodoru8** [comment]
Forgot to mention that I am also unable to connect to the database from a database viewer (DBeaver). Although I was able to the previous week. I must also add that I have not tampered with the installation of serverpod in the timespan.

**Reply by wanclem** [reply]
I'm encountering the same problem on macOS. During the initial installation and launch, everything functioned smoothly. However, upon subsequent attempts to run the server, I'm consistently met with an authentication error message

**Comment by SandPod** [comment]
Hello @bteodoru8!

> Forgot to mention that I am also unable to connect to the database from a database viewer (DBeaver).

Given that you can't connect to the database using a DB viewer, the credentials you are trying to use might not be correct.

Is there a possibility that a docker Postgres volume already exists?
You can remove any existing database volume by running the command `docker compose down -v`. Try this before starting the database.

Please also check that the credentials in `docker-compose.yaml` file match the credentials in `config/passwords.yaml`.

**Reply by bteodoru8** [reply]
Hi @SandPod, thank you very much for your answer. I have done what you said several times, even went as far as to reinstall/reset docker to no avail. I checked the credentials and everything. Tried it on both macos and windows. It works on macos, but not on windows. In Dbeaver, I have tried all ip/port combinations possible to no success either. If I try a localhost, or local ip with 8090 as port I get: 

```An error occurred while setting up the SSL connection.
An error occurred while setting up the SSL connection.
```

If I try any with port 5432, I get:

```Connection to localhost:5432 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
  Connection refused: no further information
  Connection refused: no further information
```

Made a completely fresh project just to test, did not touch anything, same thing happens. I also made a test project last week, already mentioned which worked just fine, but now it does not work anymore (connecting to docker I mean).

Here is the docker-compose file.
```
version: '3.7'

services:
  postgres:
    image: postgres:14.1
    ports:
      - '8090:5432'
    environment:
      POSTGRES_USER: postgres
      POSTGRES_DB:project
      POSTGRES_PASSWORD: "pass"
    volumes:
      - project_data:/var/lib/postgresql/data
  redis:
    image: redis:6.2.6
    ports:
      - '8091:6379'
    command: redis-server --requirepass "pass"
    environment:
      - REDIS_REPLICATION_MODE=master
volumes:
 project_data:
```

I have tried adding network_mode: "host" under postgres: , but nothing changed as well as giving the container a specific name and stuff. I am 100% sure there is something wrong with docker, but I can't find anything regarding this.

**Reply by SandPod** [reply]
>  I am 100% sure there is something wrong with docker, but I can't find anything regarding this.

Yeah, it sounds like there is something fishy going on with docker.

Maybe you could try to build this simple container: https://www.tutorialspoint.com/docker/building_web_server_docker_file.htm

Then run it directly through docker and access it. If that works you could run it from a compose file and run it. Just to try and narrow down if your network is working as expected on Windows.

**Reply by bteodoru8** [reply]
Hey @SandPod, okay so I discovered something new. I was meddling around yesterday, and it turns out that if I deactivate and activate serverpod_cli again, it sometimes lets me run the server and apply the migration once. It seems like it also does not like postgres:14.1 as I only managed to do this if I had postgres:latest as image in the docker-compose file. If the server is running in the terminal, I am also able to connect to the database from DBeaver. But it only works once, if I stop the server and try running it again, it breaks into the same loop again. In the meanwhile, same settings work completely fine on MacOS. Any ideas?

**Reply by SandPod** [reply]
Sorry, I haven't heard about this issue before.

Since the database isn't working as expected it sounds like a more general docker issue. Could you try with a general container that is not built by us to validate this?

If so, you will get better support from the Docker forum.

**Reply by bteodoru8** [reply, marked as answer]
Hey @SandPod, so I went as far as to reinstall windows earlier this week, and havent had problems since then. Truly a strange thing.

**Reply by SandPod** [reply]
I'm glad the issue got resolved and hope it doesn't surface again.

Happy coding 🚀
